### PR TITLE
make DataFrame constructor more flexible

### PR DESCRIPTION
**Pull request by bkamins**
Fixes https://github.com/JuliaData/DataFrames.jl/issues/2858

### CODE REVIEW

**Comment by bkamins** [comment]
As usual with dispatch it is more tricky than it seems, but I hope all is good 😄.

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
The CI failure is Out of Memory error - re-triggering just to make sure all is OK.

**Comment by bkamins** [comment]
@nalimilan - I have added the `copycols` kwarg support for matrix constructor. Can you please re-review the PR? Thank you!

File: src/dataframe/dataframe.jl [hunk 1]
```diff
@@ -346,22 +355,25 @@ function DataFrame(columns::AbstractVector, cnames::Symbol; copycols::Bool=true)
...
 end
 
-DataFrame(columns::AbstractMatrix, cnames::AbstractVector{Symbol}; makeunique::Bool=false) =
-    DataFrame(AbstractVector[columns[:, i] for i in 1:size(columns, 2)], cnames,
-              makeunique=makeunique, copycols=false)
+function DataFrame(columns::AbstractMatrix, cnames::AbstractVector{Symbol};
+                   makeunique::Bool=false, copycols::Bool=true)
+    getter = copycols ? getindex : view
+    return DataFrame(AbstractVector[getter(columns, :, i) for i in 1:size(columns, 2)],
+              cnames, makeunique=makeunique, copycols=false)
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl:362]:
```suggestion
    return DataFrame(AbstractVector[getter(columns, :, i) for i in 1:size(columns, 2)],
                     cnames, makeunique=makeunique, copycols=false)
```

File: test/constructors.jl [hunk 2]
```diff
@@ -372,4 +372,42 @@ end
...
+        @test df.x1 isa Vector{Int}
+        @test df.x2 isa Vector{Int}
+        df = DataFrame(m, cnames, copycols=true)
+        @test df == refdf
+        @test df.x1 isa Vector{Int}
+        @test df.x2 isa Vector{Int}
+        df = DataFrame(m, cnames, copycols=false)
+        @test df == refdf
+        @test df.x1 isa SubArray{Int,1, Matrix{Int}}
+        @test df.x2 isa SubArray{Int,1, Matrix{Int}}
```
Comment (nalimilan) [inline test/constructors.jl:407]:
```suggestion
        @test df.x1 isa SubArray{Int, 1, Matrix{Int}}
        @test df.x2 isa SubArray{Int, 1, Matrix{Int}}
```

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
OK to merge after CI passes?

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**